### PR TITLE
[avmfritz] Added support for templates

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added support for templates

Test version:
[org.openhab.binding.avmfritz-2.5.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/2702200/org.openhab.binding.avmfritz-2.5.0-SNAPSHOT.jar.zip)

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
@openhab/2-x-add-ons-maintainers May I ask for a review? Thanks.

**Comment by martinvw** [comment]
> @openhab/2-x-add-ons-maintainers May I ask for a review? Thanks.

Yes you may :-) 

Happy new year!

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseBridgeHandler.java [hunk 1]
```diff
@@ -435,10 +476,24 @@ public String getThingName(AVMFritzBaseModel device) {
...
+        logger.debug("Handle command '{}' for channel {}", command, channelId);
+        if (command == RefreshType.REFRESH) {
             handleRefreshCommand();
         }
+        FritzAhaWebInterface fritzBox = getWebInterface();
+        if (fritzBox == null) {
+            logger.debug("Cannot handle command '{}' because connection is missing", command);
+            return;
+        }
+        switch (channelId) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseBridgeHandler.java]: Imho, this is not the most sensible use for a switch statement

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java [hunk 2]
```diff
@@ -146,6 +155,12 @@ public void handleCommand(ChannelUID channelUID, Command command) {
             case CHANNEL_VOLTAGE:
                 logger.debug("Channel {} is a read-only channel and cannot handle command '{}'", channelId, command);
                 break;
+            case CHANNEL_APPLY_TEMPLATE:
+                if (command instanceof StringType) {
+                    fritzBox.applyTempalte(command.toString());
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java]: typo `applyTempalte`

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzAhaWebInterface.java [hunk 3]
```diff
@@ -180,13 +181,7 @@ protected String createResponse(String challenge) {
...
             return "";
         }
-        byte[] handshakeHash;
-        try {
-            handshakeHash = md5.digest(handshake.getBytes("UTF-16LE"));
-        } catch (UnsupportedEncodingException e) {
-            logger.error("This version of Java does not understand UTF-16LE encoding");
-            return "";
-        }
+        byte[] handshakeHash = md5.digest(handshake.getBytes(StandardCharsets.UTF_16LE));
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzAhaWebInterface.java]: 👍

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzAhaWebInterface.java [hunk 4]
```diff
@@ -295,10 +290,15 @@ public FritzAhaContentExchange asyncPost(String path, String args, FritzAhaCallb
         FritzAhaContentExchange postExchange = new FritzAhaContentExchange(callback);
         httpClient.newRequest(getURL(path)).timeout(config.getAsyncTimeout(), TimeUnit.SECONDS).method(POST)
                 .onResponseSuccess(postExchange).onResponseFailure(postExchange) // .onComplete(postExchange)
-                .content(new StringContentProvider(addSID(args), "UTF-8")).send(postExchange);
+                .content(new StringContentProvider(addSID(args), StandardCharsets.UTF_8)).send(postExchange);
         return postExchange;
     }
 
+    public FritzAhaContentExchange applyTempalte(String ain) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzAhaWebInterface.java]: Source of the typo

**Comment by cweitkamp** [comment]
@martinvw Thanks for your quick review. I incorporated the comments.

Happy new year too.

**Pull request merged**